keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,497 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP aims to introduce node-affinity-based eviction.
+In this KEP, we will introduce a new nodeAffinity type: `requiredDuringSchedulingRequiredDuringExecution`.
+A new controller called `node-affinity-eviction` will be added to the kube-controller-manager.
+This controller will monitor changes in node labels, and try to evict pods if their `requiredDuringSchedulingRequiredDuringExecution` node affinity are no longer met.
+If eviction fails, the pods that failed to be evicted will be rechecked for violations of the `requiredDuringSchedulingRequiredDuringExecution` affinity during the next node label update. 
+If, in the next eviction round, they still violate the `requiredDuringSchedulingRequiredDuringExecution` affinity, they will be attempted for eviction again.
+This controller will respect PDB during the eviction.
+
+We'll start using _the Eviction API, respecting PDB_ to evict pods during alpha stage for simplicity, 
+and may transfer to the _Evacuation API_ in future stage.
+
+## Motivation
+
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+
+- Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **nodeAffinity** type.
+- Add `node-affinity-eviction` controller to ensure pods being evicted if the selectors are no longer met at runtime, with the respect of PDB the same time.
+- DaemonSet controller will take the `RequiredDuringSchedulingRequiredDuringExecution` into consideration, 
+when recalculating the eligible nodes where the DaemonSet pods can run.
+
+### Non-Goals
+
+- This KEP does not aim to introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAffinity** type and **podAntiAffinity** type.
+
+## Proposal
+- Implement the `requiredDuringSchedulingRequiredDuringExecution` node affinity type.
+- Update the kube-scheduler, kube-controller-manager and kube-apiserver.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories 
+
+#### manage multi tenant cluster with dedicated nodes
+
+I have some nodes labeled with "userA=allow" and "userB=allow", which means that these nodes are available to user A and user B.
+As the scale grows I want to add new nodes only for user B while leaving the current nodes only for user A.
+I want to migrate all pods available to user B to new nodes, and I want to ensure HA of my services during the migration.
+
+Without `node-affinity-eviction`, I have to remove "userB=allow" label of node, and delete the pods manually. Also I can't use taints because they don't respect PDBs.
+With `node-affinity-eviction`, I can simply delete the "userB=allow" label from the existing nodes 
+to re-schedule all pods of user B to these new nodes.
+
```

Comment:
I know I'm late to the party, I just noticed this KEP is coming. 

So, I'm not sure if this user story is strong enough to support this feature, especially given implementing/maintaining a new controller is a huge cost.
Can we add some others? 

For this current scenario, why can't we just delete the pods actually (like via the deployment(s) restarts)? You're trying to argue that that's too troublesome? If you ask your ChatGPT, it'll give you a simple script to do that in a few sec.